### PR TITLE
[NEW] Mandatory 2fa for role

### LINKED ISSUE #7275: Option for mandatory Two Factor Auth for Users in Certain Roles
Similar to #7100 I'm looking for a way to **require** MFA for certain roles, like administrators. This feature could also be used to require MFA for everybody except certain roles, like bots.

I think this would be a great addition to the MFA feature introduced recently, since there is currently no good way to get adoption from users.

### PR DESCRIPTION
**Pull request by Hudell**
@RocketChat/core 

Closes #7275 

Adds a new setting on the role to determine if the user should be forced to use 2fa:

New Setting

When this setting is turned on, every user in that role that don't yet use 2fa will see this screen:

2fa Setup

### CODE REVIEW

**Comment by geekgonecrazy** [comment]
Love the idea of this.  Not sure about the presentation when they need 2fa but not enabled.  I think we need a bit cleaner way maybe?

@karlprieb @arthurgiustina thoughts?  Would a dialog maybe informing them 2FA is required be better?  That then taking them to the page, and just block at routing level to only that page until they've added?

File: packages/rocketchat-authorization/client/views/permissionsRole.html [hunk 1]
```diff
@@ -21,6 +21,11 @@
 						<option value="Users" selected="{{$eq scope 'Users'}}">{{_ "Global"}}</option>
 						<option value="Subscriptions" selected="{{$eq scope 'Subscriptions'}}">{{_ "Rooms"}}</option>
 					</select>
+
+					<br/>
+					<label for="mandatory2fa">{{_ "Users must use Two Factor Authentication"}} :</label>
```
Comment (sampaiodiego) [inline packages/rocketchat-authorization/client/views/permissionsRole.html]: we have not created a translation key for `Users must use Two Factor Authentication`

File: packages/rocketchat-ui-master/client/main.js [hunk 2]
```diff
@@ -154,6 +154,23 @@ Template.main.helpers({
...
 	},
+	require2faSetup() {
+		const user = Meteor.user();
+
+		// User is already using 2fa
+		if (user.services.totp !== undefined && user.services.totp.enabled) {
+			return false;
+		}
+
+		return user.roles.reduce((ret, role) => {
```
Comment (sampaiodiego) [inline packages/rocketchat-ui-master/client/main.js]: this `reduce` could be changed to a single `findOne`:
```
RocketChat.models.Roles.findOne({ _id: { $in: user.roles }, mandatory2fa: true }})
```

File: packages/rocketchat-authorization/client/views/permissionsRole.html (hunk 1 above)
Comment (Hudell) [reply packages/rocketchat-authorization/client/views/permissionsRole.html]: Added.

File: packages/rocketchat-ui-master/client/main.js (hunk 2 above)
Comment (Hudell) [reply packages/rocketchat-ui-master/client/main.js]: Changed.

**Comment by karlprieb** [comment]
I talk to @Hudell about that and we decided to show a modal instead this page.

He will implement this modal in the next commits.

**Comment by engelgabriel** [comment]
@Hudell can you describe the flow here to @thiagosanchz so we can finalize the UX for this PR. 

It has become a blocker for some enterprise customers.

**Comment by thiagosanchz** [comment]
@Hudell and @engelgabriel Like that? alert 2fa

**Comment by geekgonecrazy** [comment]
Only thing to note here is will need to replace the wording about mobile if you visually you expect that to be filled.

@rafaelks please correct me if i'm wrong.. But i'm 99% positive all mobile apps support 2FA and hae for a while now :)

**Comment by Hudell** [comment]
I would change the Close button for a Logout button, as logging out is the only way to get rid of that window without enabling 2FA.

**Comment by marceloschmidt** [comment]
"currently" instead of "corrently"
"Once" instead of "One"

**Comment by geekgonecrazy** [comment]
IMO should change the wording below to something like:

> Until you setup 2FA you will be not be allowed to login on any device.

Since as mentioned above Mobile **does support 2FA** and has for a long time :)

**Comment by rafaelks** [comment]
@geekgonecrazy It's been a while now, haha... but yes! Both iOS & Android native apps support 2FA.

**Comment by thiagosanchz** [comment]
@geekgonecrazy and @Hudell Since there is an important action on this alert, so I need to add a question to reinforce the decision. Wdyt? 
screen shot 2018-12-07 at 15 41 48

**Comment by geekgonecrazy** [comment]
Makes sense!  Only wording tweak would be:

> Would you like to enable now?

**Comment by thiagosanchz** [comment]
Thanks @geekgonecrazy Done!

**Comment by geekgonecrazy** [comment]
@Hudell @rodrigok did the UI designed above get implemented?  Doesn't look like any wording changes were made before merging this PR

**Comment by horstepipe** [comment]
you can easily bypass this rule by logging in via mobile. At least on IOS I can log in although the rule is enabled.

**Comment by geekgonecrazy** [comment]
@Hudell 🔝 After looking at this PR again it looks like this is only being enforced at client side level?

**Comment by sampaiodiego** [comment]
@geekgonecrazy true.. and the UI is still very bad.. :disappointed:

**Comment by jaytat0** [comment]
@rodrigok @geekgonecrazy @engelgabriel so is there actually a way to make 2fa mandatory by role - as this issue was written for - and we're missing it? We were of the impression that it got in since this was merged, but currently not seeing how to enforce it.

This is a pretty big one for us.

**Comment by Hudell** [comment]
Go to permission, modify the role you want to have mandatory 2fa and on the "Role editing" screen, there will be an option to enforce it. Please check the comments above because there are still a few problems with it.

**Comment by jaytat0** [comment]
@Hudell Ah I see. Great. Thanks for the prompt response.

**Pull request merged**